src/rsa/blinding.rs

```diff
@@ -0,0 +1,585 @@
+                sampling_params: &sampling_params,
+            };
+            range.sample_into_limbs(ext_dest, rng)
+        })
+    } else {
+        let range = Range {
+            max_exclusive: max_exclusive,
+            sampling_params: &sampling_params,
+        };
+        range.sample_into_limbs(dest, rng)
+    }
+}
+
+/// Copy `dest` and `max_exclusive` into temporary buffers which extend capacity
+/// by one, pass those to the callback, then copy the modified destination
+/// buffer back to `dest`.
+fn with_extended_buffers<F>(dest: &mut [Limb], max_exclusive: &[Limb], cb: F)
+                            -> Result<(), error::Unspecified>
+                            where F: Fn(&mut [Limb], &[Limb])
+                                     -> Result<(), error::Unspecified> {
+    const BUF_SIZE: usize = RSA_KEY_MAX_LIMBS + 1;
+
+    let buf_needed = dest.len() + 1;
+    debug_assert_eq!(buf_needed, max_exclusive.len() + 1); // equivalent definition
+
+    assert!(buf_needed <= BUF_SIZE);
+
+    let mut tmp_max: [Limb; BUF_SIZE] = [0; BUF_SIZE];
+    let mut tmp_dest: [Limb; BUF_SIZE] = [0; BUF_SIZE];
+
+    tmp_dest[..dest.len()].copy_from_slice(&dest);
+    tmp_max[..max_exclusive.len()].copy_from_slice(&max_exclusive);
+
+    let result = cb(&mut tmp_dest[..buf_needed], &tmp_max[..buf_needed]);
+
+    if result.is_ok() {
+        let dest_len = dest.len();
+        dest.copy_from_slice(&tmp_dest[..dest_len]);
+    }
+
+    result
+}
+
+/// Decide implementation strategy for random sampling.
+//
+// We support a special case performance optimization for bounds of the form
+// `0b100...` - see comment in `SamplingParams`.
+//
+// However, for simplicity, we only support this for the case when the number
+// of bits in the bound (/ public key modulus) is a multiple of LIMB_BITS,
+// or one less, which we expect to be the case in performance-sensitive
+// applications, where, e.g., the 2048 or 2047-bit modulus will be the product
+// of two 1024-bit integers.
+fn select_sampling_params(max_exclusive: &[Limb]) -> SamplingParams {
+    let most_sig = max_exclusive[max_exclusive.len() - 1];
+
+    if most_sig >> (LIMB_BITS - 3) == 0b100 {
+        SamplingParams {
+            most_sig_limb_mask: 1, // effectively a carry over into a new, more-significant limb
+            reduce_when_over_bound: true,
+            extend_limbs_by_one: true,
+        }
+    } else if most_sig >> (LIMB_BITS - 4) == 0b0100 {
+        SamplingParams {
+            most_sig_limb_mask: Limb::max_value(),
+            reduce_when_over_bound: true,
+            extend_limbs_by_one: false,
+        }
+    } else {
+        SamplingParams {
+            most_sig_limb_mask: Limb::max_value() >> most_sig.leading_zeros(),
+            reduce_when_over_bound: false,
+            extend_limbs_by_one: false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {core,rand};
+    use limb::*;
+
+    #[test]
+    fn test_select_sampling_params() {
```

Comment:
All of these test cases are one limb long, so they are not verifying that we're looking at the high limb instead of, say, the low limb.